src/easylink/step.py

```diff
@@ -401,140 +313,219 @@ def _get_step_graph(self, nodes: list[Step], edges: list[EdgeParams]) -> StepGra
             step_graph.add_edge_from_params(edge)
         return step_graph
 
+    @property
+    def implementation_node_name(self) -> str:
+        """Resolve a sensible unique node name for the implementation graph.
+        This method compares the step node names with the step names through the step hierarchy and
+        uses the full suffix of step names starting from wherever the two first differ. For example,
+        loop steps may have multiple loops with the same implementation and step, e.g. "step_3" and "step_3_python_pandas".
+        If we have node names step_3_loop_1 step_3_python_pandas the resulting implementation node name
+        will be step_3_loop_1_step_3_python_pandas. If all the node names and step names match, we have not
+        introduced any step degeneracies with e.g. loops or multiples, and we can simply use the implementation
+        name."""
+        step = self
+        node_names = []
+        step_names = []
+        while step:
+            node_names.append(step.name)
+            step_names.append(step.step_name)
+            step = step.parent_step
 
-class HierarchicalStep(CompositeStep, BasicStep):
-    """A HierarchicalStep can be a single implementation or several 'substeps'. This requires
-    a "substeps" key in the step configuration. If no substeps key is present, it will be treated as
-    a single implemented step."""
+        implementation_names = []
+        step_names.reverse()
+        node_names.reverse()
```

Comment:
These are lists. Do you need to be worried about the order in which they get appending above?